README.md
# UMich Altmetric Data
Display Altmetric data using a Wordpress shortcode or block.

You will need to generate an API URL in Altmetric Explorer. (Export Search Results -> Open Results in API) Please see Altmetric's support page for additional help: https://help.altmetric.com/support/solutions/articles/6000241368-introduction-to-the-explorer-api

### Shortcode
```
[altmetric url="{MY_URL_HERE}" limit="25" donut-size="small" template="default"]
```
OR
```
[altmetric limit="25" donut-size="small" template="default"]{MY_URL_HERE}[/altmetric]
```

#### Shortcode options
| Option     | Values        | Default |
| ---------- | ------------- | ------- |
| url        | string        |         |
| limit      | number        | 25      |
| donut-size | small, medium | medium  |
| template   | string        | default |


CHANGELOG
1.3 (2022-04-13)
==================
added Gutenberg block

1.2.5 (2021-11-17)
==================
resolve issue with spaces in search term not being encoded for api call

1.2.4 (2021-04-01)
==================
cache cleanup bugfix

1.2.3 (2020-04-07)
==================
resolves issue with reprocessing url and altmetric validation

1.2.2 (2020-01-28)
==================
resolves #2 (don't require short_open_tag)

1.2.1 (2018-09-10)
==================
minor bug fixes

1.2 (2018-04-30)
================
Add pubmed id for donut

1.1 (2018-03-15)
================
Rename Altmetrics to Altmetric

1.0 (2018-03-08)
================
initial release


umich-altmetrics.php
<?php
/**
 * Plugin Name: University of Michigan: Altmetrics
 * Plugin URI: https://github.com/umdigital/umich-altmetrics/
 * Description: Display Altmetrics attention lists via API
 * Version: 1.0
 * Author: U-M: Digital
 * Author URI: http://vpcomm.umich.edu
 */

define( 'UMICHALTMETRICS_PATH', dirname( __FILE__ ) . DIRECTORY_SEPARATOR );

class UmichAltmetrics {
    static public $displayCounter = 0;

    static private $_cacheTimeout = 6;

    static public function init()
    {
        self::$_cacheTimeout = 60 * 60 * (self::$_cacheTimeout >= 1 ?: self::$_cacheTimeout);

        // UPDATER SETUP
        if( !class_exists( 'WP_GitHub_Updater' ) ) {
            include_once UMICHALTMETRICS_PATH .'includes'. DIRECTORY_SEPARATOR .'updater.php';
        }
        if( isset( $_GET['force-check'] ) && $_GET['force-check'] && !defined( 'WP_GITHUB_FORCE_UPDATE' ) ) {
            define( 'WP_GITHUB_FORCE_UPDATE', true );
        }
        if( is_admin() ) {
            new WP_GitHub_Updater(array(
                // this is the slug of your plugin
                'slug' => plugin_basename(__FILE__),
                // this is the name of the folder your plugin lives in
                'proper_folder_name' => dirname( plugin_basename( __FILE__ ) ),
                // the github API url of your github repo
                'api_url' => 'https://api.github.com/repos/umdigital/umich-altmetrics',
                // the github raw url of your github repo
                'raw_url' => 'https://raw.githubusercontent.com/umdigital/umich-altmetrics/master',
                // the github url of your github repo
                'github_url' => 'https://github.com/umdigital/umich-altmetrics',
                 // the zip url of the github repo
                'zip_url' => 'https://github.com/umdigital/umich-altmetrics/zipball/master',
                // wether WP should check the validity of the SSL cert when getting an update, see https://github.com/jkudish/WordPress-GitHub-Plugin-Updater/issues/2 and https://github.com/jkudish/WordPress-GitHub-Plugin-Updater/issues/4 for details
                'sslverify' => true,
                // which version of WordPress does your plugin require?
                'requires' => '3.0',
                // which version of WordPress is your plugin tested up to?
                'tested' => '3.9.1',
                // which file to use as the readme for the version number
                'readme' => 'README.md',
                // Access private repositories by authorizing under Appearance > Github Updates when this example plugin is installed
                'access_token' => '',
            ));
        }

        add_action( 'wp_enqueue_scripts', function(){
            // load altmetrics donut
            wp_enqueue_script('umich-altmetrics-donut', 'https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js' );

            wp_enqueue_style( 'umich-altmetrics', plugins_url('assets/umich-altmetrics.css', __FILE__ ) );
        });

        add_shortcode( 'altmetrics', function( $atts, $url = '' ){
            self::$displayCounter++;

            $atts = shortcode_atts(array(
                'url'        => '',
                'limit'      => 25,
                'donut-size' => 'medium',
                'template'   => 'default',
            ), $atts );

            $atts['donut-size'] = $atts['donut-size'] == 'small' ? '' : $atts['donut-size'];

            $url = htmlspecialchars_decode( urldecode( trim( strip_tags( $url ) ) ) );
            if( $url && filter_var( $url, FILTER_VALIDATE_URL ) ) {
                $atts['url'] = $url;
            }

            // normalize url
            $atts['url'] = htmlspecialchars_decode( urldecode( $atts['url'] ) );

            // add limit "page[size]=#" to url
            $parts = parse_url( urldecode( $atts['url'] ) );
            parse_str( $parts['query'], $parts['query'] );
            $parts['query']['page']['size'] = $atts['limit'] ?: 1;
            $parts['query'] = http_build_query( $parts['query'] );
            $parts['query'] = $parts['query'] ? '?'. $parts['query'] : '';
            $atts['url'] = "{$parts['scheme']}://{$parts['host']}{$parts['path']}{$parts['query']}";

            // locate template
            $tpl = implode( DIRECTORY_SEPARATOR, array( UMICHALTMETRICS_PATH, 'templates', 'default.tpl' ) );
            $tpl = locate_template( array( 'umich-altmetrics/'. $atts['template'] ), false ) ?: $tpl;

            // GET ALTMETRICS DATA
            $wpUpload  = wp_upload_dir();
            $cachePath = implode( DIRECTORY_SEPARATOR, array(
                $wpUpload['basedir'],
                'umich-altmetrics-cache',
                md5( $atts['url'] ) .'.cache'
            ));

            if( !file_exists( $cachePath ) || ((@filemtime( $cachePath ) + self::$_cacheTimeout) < time()) ) {
                // update timestamp so other request don't make a pull request at the same time
                @touch( $cachePath );

                // get live results
                $stream = stream_context_create(array(
                    'http' => array(
                        'timeout' => 1
                    )
                ));
                if( $json = file_get_contents( $atts['url'], false, $stream ) ) {
                    if( $res = @json_decode( $json ) ) {
                        if( @$res->meta->response->status == 'ok' ) {
                            // CACHE RESULTS
                            wp_mkdir_p( dirname( $cachePath ) );

                            @file_put_contents( $cachePath, $json );
                        }
                    }
                }
            }

            if( $resItems = @json_decode( file_get_contents( $cachePath ) ) ) {
                $researchItems = array(
                    'items'         => array(),
                    'relationships' => array()
                );

                if( $resItems->included ) {
                    foreach( $resItems->included as $item ) {
                        $researchItems['relationships'][ $item->type ][ $item->id ] = $item->attributes;
                    }
                }

                if( $resItems->data ) {
                    foreach( $resItems->data as $item ) {
                        $researchItems['items'][ $item->id ] = $item;
                    }
                }

                // DISPLAY DATA
                ob_start();
                include( $tpl );
                return ob_get_clean();
            }
        });

        // 10% chance of cleanup
        if( mt_rand( 1, 10 ) == 3 ) {
            $expires = 60 * 60 * 24 * 7; // 7 days

            $wpUpload  = wp_upload_dir();
            $cachePath = implode( DIRECTORY_SEPARATOR, array(
                $wpUpload['basedir'],
                'umich-altmetrics-cache',
                '*'
            ));

            foreach( glob( $cachPath ) as $file ) {
                if( (filemtime( $file ) + $expires) < time() ) {
                    unlink( $file );
                }
            }
        }
    }
}
UmichAltmetrics::init();


blocks/altmetric/block.php
<?php

class UmichAltmetric_Block_Altmetric
{
    static private $_prefix = 'umichaltmetric-altmetric';
    static private $_block  = 'altmetric';

    static public function init()
    {
        $script       = null;
        $styles       = null;
        $editorStyles = null;
        $editorScript = null;

        // FRONT & BACK END JS
        if( file_exists( dirname( __FILE__ ) . DIRECTORY_SEPARATOR .'script.js' ) ) {
            $script = self::$_prefix .'--'. self::$_block .'-js';

            wp_register_script(
                $script,
                plugins_url( '/script.js', __FILE__ ),
                array( 'wp-blocks', 'wp-element', 'wp-editor', 'wp-api' ),
                filemtime( dirname( __FILE__ ) . DIRECTORY_SEPARATOR .'script.js' )
            );
        }

        // FRONT & BACKEND STYLES
        if( file_exists( dirname( __FILE__ ) . DIRECTORY_SEPARATOR .'styles.css' ) ) {
            $style = self::$_prefix .'--'. self::$_block .'-css';

            wp_register_style(
                $style,
                plugins_url( '/styles.css', __FILE__ ),
                array(),
                filemtime( dirname( __FILE__ ) . DIRECTORY_SEPARATOR .'styles.css' )
            );
        }

        // BACKEND STYLES
        if( file_exists( dirname( __FILE__ ) . DIRECTORY_SEPARATOR .'editor.css' ) ) {
            $editorStyles = self::$_prefix .'--'. self::$_block .'-ed-css';

            wp_register_style(
                $editorStyles,
                plugins_url( '/editor.css', __FILE__ ),
                array(),
                filemtime( dirname( __FILE__ ) . DIRECTORY_SEPARATOR .'editor.css' )
            );
        }

        $editorScript = self::$_prefix .'--'. self::$_block .'-ed-js';
        wp_register_script(
            $editorScript,
            plugins_url( '/editor.js', __FILE__ ),
            array( 'wp-blocks', 'wp-element', 'wp-editor', 'wp-api' ),
            filemtime( dirname( __FILE__ ) . DIRECTORY_SEPARATOR .'editor.js' )
        );

        register_block_type( __DIR__, array(
            'script'          => $script,
            'style'           => $style,
            'editor_style'    => $editorStyles,
            'editor_script'   => $editorScript,
            'render_callback' => function( $instance, $content ){
                $instance = array_merge(array(
                    'url'           => null,
                    'badgeType'     => 'medium-donut',
                    'badgePosition' => 'right',
                    'limit'         => 25,
                    'template'      => 'default',
                    'className'     => 'is-style-basic'
                ), $instance );

                $classes = array();
                $classes[] = 'wp-block-umichaltmetric-altmetric';
                $classes[] = "uma-badge-type--{$instance['badgeType']}";
                $classes[] = $instance['className'];

                $instance['className'] = implode( ' ', $classes );

                $altmetricRes = UmichAltmetric::get(array(
                    'url'   => $instance['url'],
                    'limit' => $instance['limit']
                ));

                if( $altmetricRes ) {
                    ob_start();
                    $template = implode( DIRECTORY_SEPARATOR, array( UmichAltmetric::$pluginPath, 'templates', 'block-default.tpl' ) );
                    if( $tpl = locate_template( array( "umich-altmetric/block-{$instance['template']}.tpl" ), false ) ) {
                        $template = $tpl;
                    }

                    include $template;
                    return ob_get_clean();
                }
            })
        );

        add_action( 'admin_enqueue_scripts', function(){
            wp_enqueue_script('umich-altmetric-donut', 'https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js' );
        });
    }
}
UmichAltmetric_Block_Altmetric::init();


blocks/altmetric/editor.js
(function(){
    const __ = wp.i18n.__;
    const _x = wp.i18n._x;
    const { useBlockProps, InnerBlocks } = wp.blockEditor;
    const { createElement } = wp.element;
    let altmetricTimer = false;

    wp.blocks.registerBlockType( 'umichaltmetric/altmetric', {
        edit: ( props ) => {
            const [ url, setURL ] = wp.element.useState( props.attributes.url );
            const [ isEditingURL, setIsEditingURL ] = wp.element.useState( ( props.attributes.url ? false : true ) );

            if( !isEditingURL ) {
                clearTimeout( altmetricTimer );

                altmetricTimer = setTimeout( _altmetric_embed_init, 2000 );
            }

            return createElement(
                'div',
                useBlockProps({
                    className: props.className
                }),
                ( !isEditingURL &&
                    createElement( wp.blockEditor.BlockControls,
                        null,
                        createElement( wp.components.ToolbarGroup,
                            null,
                            createElement( wp.components.ToolbarButton, {
                                    label    : __( 'Edit URL' ),
                                    className: 'components-toolbar__control',
                                    icon     : 'edit',
                                    onClick  : function(){
                                        setIsEditingURL( true );
                                    }
                                }
                            )
                        )
                    )
                ),
                createElement( wp.blockEditor.InspectorControls,
                    null,
                    createElement(
                        wp.components.PanelBody, {
                            title: 'Display Options',
                            initialOpen: true
                        },
                        createElement(
                            wp.components.SelectControl, {
                                label: 'Badge Type',
                                value: props.attributes.badgeType,
                                options: [{
                                    value: '',
                                    label: 'None'
                                },{
                                    value: 'badge',
                                    label: 'Badge'
                                },{
                                    value: 'donut',
                                    label: 'Small Donut'
                                },{
                                    value: 'medium-donut',
                                    label: 'Medium Donut'
                                },{
                                    value: 'large-donut',
                                    label: 'Large Donut'
                                },{
                                    value: 'bar',
                                    label: 'Small Bar'
                                },{
                                    value: 'medium-bar',
                                    label: 'Medium Bar'
                                },{
                                    value: 'large-bar',
                                    label: 'Large Bar'
                                }],
                                onChange: function( value ){
                                    props.setAttributes({
                                        badgeType: value
                                    });
                                },
                                help: createElement(
                                    wp.components.ExternalLink, {
                                        href: 'https://api.altmetric.com/embeds.html#badge-types'
                                    },
                                    __( 'Learn more about Altmetric Badges' )
                                )
                            }
                        ),
                        createElement(
                            wp.components.SelectControl, {
                                label: 'Badge Popover',
                                value: props.attributes.badgePopover,
                                options: [{
                                    value: '',
                                    label: 'None'
                                },{
                                    value: 'top',
                                    label: 'Top'
                                },{
                                    value: 'right',
                                    label: 'Right'
                                },{
                                    value: 'bottom',
                                    label: 'Bottom'
                                },{
                                    value: 'left',
                                    label: 'Left'
                                }],
                                onChange: function( value ){
                                    props.setAttributes({
                                        badgePopover: value
                                    });
                                },
                                help: createElement(
                                    wp.components.ExternalLink, {
                                        href: 'https://api.altmetric.com/embeds.html#popovers'
                                    },
                                    __( 'Learn more about Altmetric Badge Popovers' )
                                )
                            }
                        ),
                        createElement(
                            wp.components.TextControl, {
                                label: 'Items to show',
                                type:  'number',
                                value: props.attributes.limit,
                                onChange: function( value ){
                                    props.setAttributes({
                                        limit: parseInt( value )
                                    });
                                }
                            }
                        ),
                        createElement(
                            wp.components.TextControl, {
                                label: 'Custom Template',
                                type : 'text',
                                help : 'Custom template to use instead of the default one.',
                                value: props.attributes.template,
                                onChange: function( value ){
                                    props.setAttributes({
                                        template: value
                                    });
                                }
                            }
                        )
                    )
                ),
                (
                    isEditingURL &&
                    createElement(
                        wp.components.Placeholder, {
                            icon: 'admin-site-alt3',
                            label: __('Altmetric API URL'),
                            instructions: __( 'URL can be found at altmetric.com, then for the desired report go to Research Outputs Tab -> Export This Tab -> Open results in API' )
                        },
                        createElement(
                            'form', {
                                onSubmit: function( event ){
                                    if( event ) {
                                        event.preventDefault();
                                    }

                                    setIsEditingURL( false );
                                    props.setAttributes({
                                        url: url
                                    });
                                }
                            },
                            createElement(
                                'input', {
                                    type: 'url',
                                    value: url,
                                    className: 'components-placeholder__input',
                                    'aria-label': 'URL',
                                    placeholder: __( 'Enter URL to embed here…' ),
                                    onChange: function( event ) {
                                        setURL( event.target.value );
                                    }
                                }
                            ),
                            createElement(
                                'button', {
                                    type: 'submit',
                                    className: 'components-button is-primary'
                                },
                                _x( 'Embed', 'button label' )
                            )
                        ),
                        createElement(
                            'div', {
                                className: 'components-placeholder__learn-more'
                            },
                            createElement(
                                wp.components.ExternalLink, {
                                    href: 'https://help.altmetric.com/support/solutions/articles/6000241368-introduction-to-the-explorer-api'
                                },
                                __( 'Learn more about Altmetric Explorer API' )
                            )
                        )
                    )
                ),
                (
                    !isEditingURL &&
                    (
                        createElement( wp.serverSideRender, {
                            block: 'umichaltmetric/altmetric',
                            attributes: props.attributes,
                        })
                    )
                )
            )
        }
    });
}());


umich-altmetric.php
<?php
/**
 * Plugin Name: University of Michigan: Altmetric
 * Plugin URI: https://github.com/umdigital/umich-altmetric/
 * Description: Display Altmetric attention lists via API
 * Version: 1.3
 * Author: U-M: Digital
 * Author URI: https://vpcomm.umich.edu
 */

class UmichAltmetric
{
    static public $pluginPath;
    static public $displayCounter = 0;

    static private $_cacheTimeout = 6;

    static public function init()
    {
        self::$pluginPath    = dirname( __FILE__ ) . DIRECTORY_SEPARATOR;
        self::$_cacheTimeout = 60 * 60 * (self::$_cacheTimeout >= 1 ? self::$_cacheTimeout : 1);

        // UPDATER SETUP
        if( !class_exists( 'WP_GitHub_Updater' ) ) {
            include_once self::$pluginPath .'includes'. DIRECTORY_SEPARATOR .'updater.php';
        }
        if( isset( $_GET['force-check'] ) && $_GET['force-check'] && !defined( 'WP_GITHUB_FORCE_UPDATE' ) ) {
            define( 'WP_GITHUB_FORCE_UPDATE', true );
        }
        if( is_admin() ) {
            new WP_GitHub_Updater(array(
                // this is the slug of your plugin
                'slug' => plugin_basename(__FILE__),
                // this is the name of the folder your plugin lives in
                'proper_folder_name' => dirname( plugin_basename( __FILE__ ) ),
                // the github API url of your github repo
                'api_url' => 'https://api.github.com/repos/umdigital/umich-altmetric',
                // the github raw url of your github repo
                'raw_url' => 'https://raw.githubusercontent.com/umdigital/umich-altmetric/master',
                // the github url of your github repo
                'github_url' => 'https://github.com/umdigital/umich-altmetric',
                 // the zip url of the github repo
                'zip_url' => 'https://github.com/umdigital/umich-altmetric/zipball/master',
                // wether WP should check the validity of the SSL cert when getting an update, see https://github.com/jkudish/WordPress-GitHub-Plugin-Updater/issues/2 and https://github.com/jkudish/WordPress-GitHub-Plugin-Updater/issues/4 for details
                'sslverify' => true,
                // which version of WordPress does your plugin require?
                'requires' => '3.0',
                // which version of WordPress is your plugin tested up to?
                'tested' => '3.9.1',
                // which file to use as the readme for the version number
                'readme' => 'README.md',
                // Access private repositories by authorizing under Appearance > Github Updates when this example plugin is installed
                'access_token' => '',
            ));
        }

        add_action( 'wp_enqueue_scripts', function(){
            // load altmetric donut
            wp_enqueue_script('umich-altmetric-donut', 'https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js' );

            wp_enqueue_style( 'umich-altmetric', plugins_url('assets/umich-altmetric.css', __FILE__ ) );
        });

        // ADD EDITOR BLOCKS
        add_action( 'init', function(){
            if( function_exists( 'register_block_type' ) ) {
                foreach( glob( __DIR__ .'/blocks/*',  GLOB_ONLYDIR ) as $block ) {
                    if( is_file( "{$block}/block.php" ) ) {
                        include_once "{$block}/block.php";
                    }
                }
            }
        });

        add_shortcode( 'altmetric', function( $atts, $url = '' ){
            self::$displayCounter++;

            $atts = shortcode_atts(array(
                'url'        => '',
                'limit'      => 25,
                'donut-size' => 'medium',
                'template'   => 'default',
            ), $atts );

            $atts['donut-size'] = $atts['donut-size'] == 'small' ? '' : $atts['donut-size'];

            $url = htmlspecialchars_decode( urldecode( trim( strip_tags( $url ) ) ) );
            if( $url && filter_var( $url, FILTER_VALIDATE_URL ) ) {
                $atts['url'] = $url;
            }

            // normalize url
            $atts['url'] = htmlspecialchars_decode( urldecode( $atts['url'] ) );

            $researchItems = self::get(array(
                'url'   => $atts['url'],
                'limit' => $atts['limit']
            ));

            // locate template
            $tpl = implode( DIRECTORY_SEPARATOR, array( self::$pluginPath, 'templates', 'default.tpl' ) );
            $tpl = locate_template( array( 'umich-altmetric/'. $atts['template'] .'.tpl' ), false ) ?: $tpl;

            // DISPLAY DATA
            if( $researchItems ) {
                ob_start();
                include( $tpl );
                return ob_get_clean();
            }
        });

        // 10% chance of cleanup
        if( mt_rand( 1, 10 ) == 3 ) {
            $expires = 60 * 60 * 24 * 7; // 7 days

            $wpUpload  = wp_upload_dir();
            $cachePath = implode( DIRECTORY_SEPARATOR, array(
                $wpUpload['basedir'],
                'umich-altmetric-cache',
                '*'
            ));

            foreach( glob( $cachePath ) as $file ) {
                if( (filemtime( $file ) + $expires) < time() ) {
                    unlink( $file );
                }
            }
        }
    }

    static public function get( $params )
    {
        $params = array_merge(array(
            'url'   => '',
            'limit' => 25
        ), $params );

        // normalize & validate url
        $params['url'] = htmlspecialchars_decode( urldecode( $params['url'] ) );
        if( !$params['url'] || !filter_var( $params['url'], FILTER_VALIDATE_URL ) ) {
            return false;
        }

        // add limit "page[size]=#" to url
        $parts = parse_url( urldecode( $params['url'] ) );
        parse_str( $parts['query'], $parts['query'] );
        $parts['query']['page']['size'] = $params['limit'] ?: 1;
        $parts['query'] = http_build_query( $parts['query'] );
        $parts['query'] = $parts['query'] ? '?'. $parts['query'] : '';
        $params['url'] = "{$parts['scheme']}://{$parts['host']}{$parts['path']}{$parts['query']}";
        $params['url'] = preg_replace( '/\[[0-9]+\]/', '[]', urldecode( $params['url'] ) );
        $params['url'] = str_replace( ' ', '%20', $params['url'] ); // spaces have to be encoded

        // GET ALTMETRICS DATA
        $wpUpload  = wp_upload_dir();
        $cachePath = implode( DIRECTORY_SEPARATOR, array(
            $wpUpload['basedir'],
            'umich-altmetric-cache',
            md5( $params['url'] ) .'.cache'
        ));

        if( !file_exists( $cachePath ) || ((@filemtime( $cachePath ) + self::$_cacheTimeout) < time()) ) {
            // update timestamp so other request don't make a pull request at the same time
            @touch( $cachePath );

            // get live results (use curl if available)
            if( function_exists( 'curl_init' ) ) {
                $curl = curl_init();
                curl_setopt( $curl, CURLOPT_URL, $params['url'] );
                curl_setopt( $curl, CURLOPT_HEADER, false );
                curl_setopt( $curl, CURLOPT_RETURNTRANSFER, true );
                curl_setopt( $curl, CURLOPT_FOLLOWLOCATION, true );
                curl_setopt( $curl, CURLOPT_CONNECTTIMEOUT, 3 ); 
                curl_setopt( $curl, CURLOPT_TIMEOUT, 10 );

                $json = curl_exec( $curl );
                curl_close( $curl );
            }
            else {
                $stream = stream_context_create(array(
                    'http' => array(
                        'timeout' => 10
                    )
                ));
                $json = file_get_contents( $params['url'], false, $stream );
            }

            if( $json && ($res = @json_decode( $json )) ) {
                if( @$res->meta->response->status == 'ok' ) {
                    // CACHE RESULTS
                    wp_mkdir_p( dirname( $cachePath ) );

                    @file_put_contents( $cachePath, $json );
                }
            }
        }

        if( $resItems = @json_decode( file_get_contents( $cachePath ) ) ) {
            $researchItems = array(
                'items'         => array(),
                'relationships' => array()
            );

            if( $resItems->included ) {
                foreach( $resItems->included as $item ) {
                    $researchItems['relationships'][ $item->type ][ $item->id ] = $item->attributes;
                }
            }

            if( $resItems->data ) {
                foreach( $resItems->data as $item ) {
                    $researchItems['items'][ $item->id ] = $item;
                }
            }

            return $researchItems;
        }

        return false;
    }
}
UmichAltmetric::init();
